### ISSUE TITLE
TRL and BERT2GPT2 model issue

### ISSUE THREAD

**Issue by lvwerra** [lvwerra/trl#24]
Hi Ayham, I migrated your question here since it is more about `trl` than `transformers`:

---
**Original post:**

Dear Werra and Blagojevic,
First of all, I would like to thank you for the great package, trl, and the fantastic notebook,
Welcome to Transformer Reinforcement Learning (trl) | trl, which explained everything about this package.

I’ve built different models using a GPT-2 decoder for the summarization task with good Rouge scores: Bert2GPT2, Roberta2GPT2, XLNet2GPT2, and Distilbert2GPT2. Then I tried to apply trl on these models to enhance the novelty of the results (producing novel words in the output summary that didn’t appear in the input text).
The input text is supposed to be the query, and the generated summary is the response.

I’ve used different functions and models for the reward:

1- reward_function1_novelty: This is a novelty function which returns the number of novel words that did not appear in the input text. This function takes ratios of single-novel-words, pair-novel-words, triple-novel-words, and four-novel-words and then returns the resulted value.

2- reward_function2_similarity: This function uses the SentenceTransformer(‘bert-base-nli-mean-tokens’) model which returns the degree of similarity between two sentences. This score can be used as a novelty reward.

3- reward_function3_paraphrasing: This function uses the AutoModelForSequenceClassification.from_pretrained(“bert-base-cased-finetuned-mrpc”) model, with its tokenizer, which returns the paraphrasing score between two sentences. This score can also help as a novelty reward.

Unfortunately, after training the models with these rewards, none of them give acceptable results in both Rouge and Novelty scores. I don’t really know where the problem is, I spent about 4 months working on this code and trying different ways to fix it but I gave up. So I would be very grateful if you could help me find the problem(s) and its solution.

The code is available here 1

Best Regards,
Ayham

**Comment by lvwerra** [comment]
The main issue I see is that you are technically not using a decoder only architecture (like GPT-2) which is the only architecture properly implemented in `trl`. Instead you use encoder-decoder architectures combining an encoder (BERT, Roberta etc.) with the GPT-2 decoder. I know that @vblagoje has worked on such architectures so maybe he has some insights in getting them to work.

Alternatively, I suggest trying to frame your summarization task a language modeling objective suitable for a GPT-2 model architecture. You could for example concatenate all texts from the summarization dataset to the following form:
```
FULL TEXT:
<here goes the long text>

SUMMARY:
<here goes the summary>

<EOS_TOKEN>
FULL TEXT:
...
```
You can then run the [language modeling script](https://github.com/huggingface/transformers/tree/master/examples/pytorch/language-modeling) with GPT-2 on this file. With that fine-tuned model you can then use `trl` since it uses only a decoder. 

Hope this helps!
